### PR TITLE
Migrate FBXReader to use model::Mesh and clean Model Part rendering

### PR DESCRIPTION
**Pull request by samcake**
- FBXReader now generate a model::Mesh object for each FBXMesh. This model::Mesh is ready for rendering and as a consequence the NetworkMesh doesn't need to allocate the vertex and index buffers anymore.

- A mesh can contains several Parts which can each have different Material. The parts are defined as indices range in the index buffer, thus when rendering the parts of  a mesh the index buffer or Vertex buffer doesn't need to change, it s just a new drawcall with different indexStart and number of indices.
 The next step is to try to combine all the FBXMeshes of a model ( and then of the models of the scene) into a single VertexBuffer and Index Buffer so we can draw it using MultiDrawIndirect in only a few drawcalls... to be continued.

- The piece of code that was generating the Merged Triangles and Quad triangles of a MeshPart has been replaced by creating the "quadTriangleIndices" in the FBXMeshPart  and using it during the model::Mesh creation.

- Fixed a bug in GLBackend::do_drawIndexed and do_drawIndexedInstanced, the offset computation for the index buffer was wrong, this is now fixed and let us draw successfully the different parts of a single Mesh.
  Because of that change we fixed the instanced shape drawing indexOffset which was passed as a byte offset and not an index.

- Moved the MeshPartPayload class and everything about rendering the Model(.h, .cpp) into the new files called ModelRender(.h, .cpp) and MeshPartPayload(.h, .cpp)
 The MeshPartPayload is now cashing the Mesh, Part and Material it s using for rendering and exposes new methods which together are contributing to the rendering of the part.

- Moved the skinning code used in the shader in it's on shared shader file "Skinning.slh" used by the 3 current skinning vector shaders.

- Lastly, i couldn't resist and replaced the old school mat4 array uniforms used to pass the skeleton current bind-pose matrices for every clusters by a UniformBuffer managed from Model.cpp and updated along the sys mem version of the clusterMatrices data.

- Breaking news, we just fixed the shiny objects issue by actually using the specular map of the model...

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4270b83c0f09f5d178c538b4a9cd4998113e6537 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fc7f0c77a5e06f82d773e728c2e4681d87a445bd available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 59afbf1a04ba3c19bb78c518b9748ece6b2d4e5c available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by ZappoMan** [comment]
Coding standard: class implementation should be in file of same name.

**Comment by murillodigital** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 59afbf1a04ba3c19bb78c518b9748ece6b2d4e5c available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by RyanDowne** [comment]
What's the best way to test this? What specifically should I be looking for? 

One thing I noticed right off the bat is that the bounding boxes on models/entities when using edit.js are incorrect. That is not true in the production build.

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+#include "model_specular_map_frag.h"
+#include "model_lightmap_frag.h"
+#include "model_lightmap_normal_map_frag.h"
+#include "model_lightmap_normal_specular_map_frag.h"
+#include "model_lightmap_specular_map_frag.h"
+#include "model_translucent_frag.h"
+
+ModelRender::RenderPipelineLib ModelRender::_renderPipelineLib;
+
+const ModelRender::RenderPipelineLib& ModelRender::getRenderPipelineLib() {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in ModelRender.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+        _renderPipelineLib.addRenderPipeline(
+            RenderKey(RenderKey::IS_SKINNED | RenderKey::IS_DEPTH_ONLY | RenderKey::IS_SHADOW),
+            skinModelShadowVertex, modelShadowPixel);
+    }
+
+    return _renderPipelineLib;
+}
+
+
+void ModelRender::RenderPipelineLib::addRenderPipeline(ModelRender::RenderKey key,
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in ModelRender.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+
+            // create a new RenderPipeline with the same shader side and the mirrorState
+            auto wireframePipeline = gpu::PipelinePointer(gpu::Pipeline::create(program, wireframeState));
+            insert(value_type(wireframeKey.getRaw(), RenderPipeline(wireframePipeline, locations)));
+        }
+    }
+}
+
+
+void ModelRender::RenderPipelineLib::initLocations(gpu::ShaderPointer& program, ModelRender::Locations& locations) {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in ModelRender.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+    locations.normalTextureUnit = program->getTextures().findLocation("normalMap");
+    locations.specularTextureUnit = program->getTextures().findLocation("specularMap");
+    locations.emissiveTextureUnit = program->getTextures().findLocation("emissiveMap");
+    locations.materialBufferUnit = program->getBuffers().findLocation("materialBuffer");
+    locations.lightBufferUnit = program->getBuffers().findLocation("lightBuffer");
+    locations.skinClusterBufferUnit = program->getBuffers().findLocation("skinClusterBuffer");
+}
+
+
+void ModelRender::pickPrograms(gpu::Batch& batch, RenderArgs::RenderMode mode, bool translucent, float alphaThreshold,
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in ModelRender.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+        return render::Item::Bound();
+    }
+    template <> void payloadRender(const MeshPartPayload::Pointer& payload, RenderArgs* args) {
+        return payload->render(args);
+    }
+}
+
+using namespace render;
+
+MeshPartPayload::MeshPartPayload(Model* model, int meshIndex, int partIndex, int shapeIndex) :
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in MeshPartPayload.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 6]
```diff
@@ -0,0 +1,638 @@
...
+    model(model), url(model->getURL()), meshIndex(meshIndex), partIndex(partIndex), _shapeID(shapeIndex)
+{
+    initCache();
+}
+
+void MeshPartPayload::initCache() {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in MeshPartPayload.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+    _drawPart = _drawMesh->getPartBuffer().get<model::Mesh::Part>(partIndex);
+
+    auto networkMaterial = model->_geometry->getShapeMaterial(_shapeID);
+    if (networkMaterial) {
+        _drawMaterial = networkMaterial->_material;
+    };
+
+}
+
+render::ItemKey MeshPartPayload::getKey() const {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in MeshPartPayload.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+        auto matKey = _drawMaterial->getKey();
+        if (matKey.isTransparent() || matKey.isTransparentMap()) {
+            builder.withTransparent();
+        }
+    }
+
+    return builder.build();
+}
+
+render::Item::Bound MeshPartPayload::getBound() const {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in MeshPartPayload.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 9]
```diff
@@ -0,0 +1,638 @@
...
+    if (_isBoundInvalid) {
+        model->getPartBounds(meshIndex, partIndex);
+        _isBoundInvalid = false;
+    }
+    return _bound;
+}
+
+void MeshPartPayload::drawCall(gpu::Batch& batch) const {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in MeshPartPayload.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 10]
```diff
@@ -0,0 +1,638 @@
...
+    batch.drawIndexed(gpu::TRIANGLES, _drawPart._numIndices, _drawPart._startIndex);
+}
+
+void MeshPartPayload::bindMesh(gpu::Batch& batch) const {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in MeshPartPayload.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 11]
```diff
@@ -0,0 +1,638 @@
...
+        batch.setInputStream(2, _drawMesh->getVertexStream().makeRangedStream(2));
+    }
+    
+    // TODO: Get rid of that extra call
+    if (!_hasColorAttrib) {
+        batch._glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
+    }
+}
+
+void MeshPartPayload::bindMaterial(gpu::Batch& batch, const ModelRender::Locations* locations) const {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in MeshPartPayload.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 12]
```diff
@@ -0,0 +1,638 @@
...
+        batch.setResourceTexture(ModelRender::LIGHTMAP_MAP_SLOT, nullptr);
+    }
+
+    // Texcoord transforms ?
+    if (locations->texcoordMatrices >= 0) {
+        batch._glUniformMatrix4fv(locations->texcoordMatrices, 2, false, (const float*)&texcoordTransform);
+    }
+}
+
+void MeshPartPayload::bindTransform(gpu::Batch& batch, const ModelRender::Locations* locations) const {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in MeshPartPayload.cpp

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 13]
```diff
@@ -0,0 +1,638 @@
...
+        } else {
+            transform = Transform(state.clusterMatrices[0]);
+        }
+    }
+    transform.preTranslate(model->_translation);
+    batch.setModelTransform(transform);
+}
+
+
+void MeshPartPayload::render(RenderArgs* args) const {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: these should be in MeshPartPayload.cpp

File: libraries/fbx/src/FBXReader_Mesh.cpp [hunk 14]
```diff
@@ -469,43 +502,48 @@ void FBXReader::buildModelMesh(ExtractedMesh& extracted, const QString& url) {
...
-        if (quadPart._numIndices) {
-            parts.push_back(quadPart);
-            ib->setSubData(offset, part.quadIndices.size() * sizeof(int),
-                           (gpu::Byte*) part.quadIndices.constData());
-            offset += part.quadIndices.size() * sizeof(int);
-            indexNum += part.quadIndices.size();
+        model::Mesh::Part modelPart(indexNum, 0, 0, model::Mesh::TRIANGLES);
+        
+        if (part.quadTrianglesIndices.size()) {
+            ib->setSubData( offset,
```
Comment (ZappoMan) [inline libraries/fbx/src/FBXReader_Mesh.cpp]: coding standard: variable names should be descriptive... what is ib? I assume it's index buffer, name it as such

File: libraries/fbx/src/FBXReader.h [hunk 15]
```diff
@@ -194,7 +189,7 @@ class FBXMesh {
 
     unsigned int meshIndex; // the order the meshes appeared in the object file
 #   if USE_MODEL_MESH
```
Comment (ZappoMan) [inline libraries/fbx/src/FBXReader.h]: is USE_MODEL_MESH defined or not? I think your goal is to have it defined... so, maybe we kill this ifdef and only have the enabled code?

File: libraries/model-networking/src/model-networking/ModelCache.h [hunk 16]
```diff
@@ -191,15 +191,14 @@ class NetworkMaterial {
 /// The state associated with a single mesh.
 class NetworkMesh {
 public:
-    gpu::BufferPointer _indexBuffer;
+    model::MeshPointer _mesh;
+
+/*    gpu::BufferPointer _indexBuffer;
```
Comment (ZappoMan) [inline libraries/model-networking/src/model-networking/ModelCache.h]: dead code? Let's kill it.. or comment why you're keeping it

**Comment by samcake** [comment]
@RyanDowne THis should just render clean and possibly a bit faster

The bounding boxes are normally not affected but i ll look

File: libraries/render-utils/src/Model.h [hunk 17]
```diff
@@ -494,6 +365,7 @@ class Model : public QObject {
     bool _needsReload = true;
     bool _needsUpdateClusterMatrices = true;
 
+    friend class MeshPartPayload;
```
Comment (ZappoMan) [inline libraries/render-utils/src/Model.h]: hmmmm.... what's this? Why do we need to be friends? this seems hacky...

**Comment by RyanDowne** [comment]
image

I see this billboard when I turn my head. It looks to be the starfield

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 18]
```diff
@@ -0,0 +1,638 @@
...
+    initLocations(program, *locations);
+
+
+    auto state = std::make_shared<gpu::State>();
+
+    // Backface on shadow
+    if (key.isShadow()) {
+        state->setCullMode(gpu::State::CULL_FRONT);
+        state->setDepthBias(1.0f);
+        state->setDepthBiasSlopeScale(4.0f);
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: magic numbers... what's 4.0???

File: libraries/render-utils/src/ModelRenderItem.cpp [hunk 19]
```diff
@@ -0,0 +1,638 @@
...
+    }
+
+    gpu::ShaderPointer program = (*pipeline).second._pipeline->getProgram();
+    locations = (*pipeline).second._locations.get();
+
+
+    // Setup the One pipeline
+    batch.setPipeline((*pipeline).second._pipeline);
+
+    if ((locations->alphaThreshold > -1) && (mode != RenderArgs::SHADOW_RENDER_MODE)) {
```
Comment (ZappoMan) [inline libraries/render-utils/src/ModelRenderItem.cpp]: coding standard: magic number? why -1? maybe DEFAULT_ALPHA_THRESHOLD or UNSET_ALPHA_THRESHOLD

**Comment by hifi-gustavo** [comment]
Build console output for commit: c92a6d58ab0d3d97f373683906d702e0d4cbd807 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

File: libraries/render-utils/src/Model.h (hunk 17 above)
Comment (samcake) [inline libraries/render-utils/src/Model.h]: We still to need easy access to the Model from the MeshPartPayload render call especially for accessing the transform information, skeleton and blendshapes information. 
They are definitely splitting, but MeshPartPayload still has the key to the Model's apartment...

**Comment by hifi-gustavo** [comment]
Build console output for commit: 429d82e6502b85bf99edb7d06931ac8ae6683440 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 978434f15ccf30e4e161d16d243fa5a8dc838b24 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1788b7b019b0217508e2cdc9488e6427e44a20c1 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: cf3828629e7f3c7426f0ab0a9af567f1ddec2418 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by samcake** [comment]
This should be working correctly normally, i discovered that some code from last night got lost when merging this morning on my office machine...

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2b5742b2778170cb809ccb646d9a6773e9bd9217 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by murillodigital** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2b5742b2778170cb809ccb646d9a6773e9bd9217 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by samcake** [comment]
Fixed the wire box issue that showed up in edit.js

**Comment by hifi-gustavo** [comment]
Build console output for commit: c75c31ddc84cbaae22af512274c7c19c66499346 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 83a0091e47f1f562a48fd8b364b288392817ab89 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dd2aa36a119293baa17ad8b027781a25def23f57 available at: https://gist.github.com/50b73c5690b975beb393<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6000.dmg

**Pull request merged**